feature(scheduler): Customizable pod selection and ordering in DefaultPreemption plugin

nickbp opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:

The current plugin behavior is to select only based on pod priority, where any pod with higher priority can preempt any pod with lower priority. If this isn't desirable, there are two options: Reimplement this plugin from scratch, or update other users of pod priority (like scheduler queue ordering) to use something else. This introduces a way to customize the selection and ordering without needing to resort to that.

In our case, we have multiple priority classes but where only a subset of those are considered preemptible. Here's roughly how this looks:

- High priority: higher in the scheduling queue, not preemptible
- Low priority: lower in the scheduling queue, not preemptible
- Preemptible priority: lowest in the scheduling queue, preemptible

This PR allows the preemption selection to be configured against the `DefaultPreemption` plugin, rather than needing to reimplement the plugin itself. The structure used here mimics [what's currently being done for the `Evaluator`](https://github.com/kubernetes/kubernetes/blob/release-1.32/pkg/scheduler/framework/preemption/preemption.go#L161-L165), where a `PreemptPod` callback may be overridden to customize what happens when performing a preemption.

This PR also updates the plugin's tests to call the updated `New()` function rather than constructing `DefaultPreemption` directly, so that `New()` is now being exercised in tests.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

nickbp committed:
feature(scheduler): Custom pod selection/ordering in DefaultPreemption

The current behavior is to select only based on pod priority, where any
pod with higher priority can preempt any pod with lower priority.

In our case, we have multiple priority classes but where only a subset
of those are considered preemptible. Here's roughly how this looks:

- High priority: higher in the scheduling queue, not preemptible
- Low priority: lower in the scheduling queue, not preemptible
- Preemptible priority: lowest in the scheduling queue, preemptible

This PR allows the preemption selection to be configured against the
`DefaultPreemption` plugin, rather than needing to reimplement the
plugin itself. The structure used here mimics [what's currently being
done for the `Evaluator`](https://github.com/kubernetes/kubernetes/blob/release-1.32/pkg/scheduler/framework/preemption/preemption.go#L161-L165),
where a `PreemptPod` callback may be overridden to customize what
happens when performing a preemption.

This PR also updates the plugin's tests to call the updated `New()`
function rather than constructing `DefaultPreemption` directly, so
that `New()` is now being exercised in tests.

haosdent commented:
> High priority: higher in the scheduling queue, not preemptible
> Low priority: lower in the scheduling queue, not preemptible

What if you set them as the same priority, and then they could not be preempted each other?

nickbp commented:
> > High priority: higher in the scheduling queue, not preemptible
> > Low priority: lower in the scheduling queue, not preemptible
> 
> What if you set them as the same priority, and then they could not be preempted each other?

The pod priority is also used by [the PrioritySort plugin](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L42-L46) for the scheduling queue, so flattening the priority structure would lose that ordering as well.

sanposhiho commented:
Looks like the same motivation as the preemptiontoleration plugin at our sub project, doesn't it?
https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/pkg%2Fpreemptiontoleration%2FREADME.md

Although I, either way, agree to put this change since we can simplify the preemptiontoleration plugin implementation.

But, I'm not sure if it requires KEP because it's no API addition, a simple change, and no impact on the default scheduler's behavior.

googs1025 commented:
In addition to pdb, PriorityClass can also be used to control which Pods can be preempted. 

by the way: we have another issue to track: https://github.com/kubernetes-sigs/scheduler-plugins/issues/839

sanposhiho commented:
Make an API (at PDB? PriorityClass?) for this kind of behavior is another story, which we can discuss separately. We need to see if there is enough demand for such behavior before proceeding for that, especially because we already have subproject and people can still just use that now.

For now, I believe we can just go ahead with this PR, that is, without KEP, and we can create a separate issue to see the reaction from users.

haosdent commented:
/ok-to-test

haosdent commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -433,13 +431,14 @@ func TestPostFilter(t *testing.T) {
+				t.Fatal(err)
+			}
+			p, ok := pi.(*DefaultPreemption)
+			if !ok {
```
Do we need a test case to verify that customized `EligiblePods` and `OrderedPods` (different from the default one) work as expected?

nickbp replied:
Yes that's a good point, I'll write up some tests that exercise customizing these.

googs1025 commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -54,6 +54,14 @@ type DefaultPreemption struct {
+
+	// EligiblePods returns victim pods which are allowed to be preempted by the provided preemptor.
+	// The default behavior is to allow any pods of lower priority to be preempted by any pods of higher priority.
+	EligiblePods func(nodeInfo *framework.NodeInfo, preemptor *v1.Pod) []*framework.PodInfo
```
If I understand correctly, from what I've seen so far, this PR doesn't seem to introduce any changes compared to the original logic. If that's the case, I'm curious how does it provide the 'Customizable' feature?  🤔

haosdent replied:
Pass the custom functions when creating `DefaultPreemption{}`?

nickbp replied:
Right, the idea is that someone would be able to assign their own functions when constructing the plugin. This should be easier to see once I've written some unit tests to exercise it.

googs1025 replied:
I am not sure if this has limitations, because we can only use this when working directly with  `DefaultPreemption{}`. It cannot be used after calling New(...) since New returns an interface.  🤔

AxeZhan replied:
That's not customizable for me. 
My thought is there should be some configuration for users to set, like `NodeResourcesFitArgs` for example. 
And that's why I said maybe we need a KEP at first🤔.
@sanposhiho What do you think? Can we create a new configuration without a KEP?

sanposhiho replied:
It's not customization for general users who want to use a default scheduler though, this is also customization just for a different layer of users. Here, the motivation comes from a user who wants to create a custom plugin, but with some customization. So, this is the same level of extensibility as [`Evaluator.Interface`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#L106-L107) that allows people to extend some behaviour from the default one.

Again, I'm not saying we shouldn't add it to the scheduler config or whatever APIs. But let's consider that separately and see if there's a big demand from the community [like I said](https://github.com/kubernetes/kubernetes/pull/129983#issuecomment-2635658180).

sanposhiho replied:
>  It cannot be used after calling New(...) since New returns an interface

You can still do type assertions. But, I'm open to a discussion how to make it extensible in a better way.

Given, like I said, we already have [`Evaluator.Interface`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#L106-L107) for an extension, the alternative option would be having other funcs there, like:

```go
type Interface interface {
...

// SelectVictimsOnNode finds...
//...
// Evaluator passes Interface.EligiblePodsFunc to eligiblePodsFunc.
  SelectVictimsOnNode(ctx context.Context, state *framework.CycleState,
		pod *v1.Pod, nodeInfo *framework.NodeInfo, pdbs []*policy.PodDisruptionBudget, 
		eligiblePodsFunc(nodeInfo *framework.NodeInfo, preemptor *v1.Pod) []*framework.PodInfo) ([]*v1.Pod, int, *framework.Status)
  EligiblePodsFunc(nodeInfo *framework.NodeInfo, preemptor *v1.Pod) []*framework.PodInfo
}
```

or maybe we can just put EligiblePodsFunc and OrderPods to `Evaluator` struct and add params to NewEvaluator.

AxeZhan replied:
> It's not customization for general users who want to use a default scheduler though, this is also customization just for a different layer of users.

In that case, I tend to we extend [Evaluator.Interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#L106-L107) here.

nickbp replied:
As a sort of workaround here, I just recently added [a call](https://github.com/kubernetes/kubernetes/pull/129983/files#diff-06607f15768495cdfff1d66f563e0c7ca8b761f6ec3a3b4b4e6d3c8b6c4cc2b9R94) to construct a `DefaultPreemption` directly, so now there's:

- `New()`: Returns `framework.Plugin`, aligns with `PluginFactoryWithFts` (no change)
- `NewDefaultPreemption()`: Same arguments as `New()`, but returns `DefaultPreemption` without needing a type assertion.

How does this look?

googs1025 replied:
> 
> or maybe we can just put EligiblePodsFunc and OrderPods to `Evaluator` struct and add params to NewEvaluator.

If a new interface method is added, it may also be necessary to consider the possibility of destroying the original interface. 🤔  If the original logic is not changed, perhaps changing the interface will do more harm than good. This is just my opinion

sanposhiho replied:
> it may also be necessary to consider the possibility of destroying 

Yeah, good point. So, we should avoid adding new funcs there so that we don't break user's implementation unnecessarily.

> As a sort of workaround here, I just recently added [a call](https://github.com/kubernetes/kubernetes/pull/129983/files#diff-06607f15768495cdfff1d66f563e0c7ca8b761f6ec3a3b4b4e6d3c8b6c4cc2b9R94) to construct a DefaultPreemption directly,

I missed this change. The current one looks the best for me over the other ones that I mentioned above. This way makes it easier for users to replace the existing `DefaultPreemption.Evaluator` as well.

AxeZhan commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -206,11 +227,11 @@ func (pl *DefaultPreemption) SelectVictimsOnNode(
 		return nil, 0, status
 	}
-	var victims []*v1.Pod
+	var victims []*framework.PodInfo
```
Why not keep the current implementation `[]*v1.Pod` ?

nickbp replied:
I had this thinking:
- Further down here, the full PodInfo is being used for `removePod()` which then calls `fh.RunPrefilterExtensionRemovePod()`, so mapping to Pods and then back to PodInfos would be tricky.
- The provided OrderedPods function could have a use for other information in the PodInfo anyway

AxeZhan commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -54,6 +54,14 @@ type DefaultPreemption struct {
+
+	// OrderedPods sorts eligible victims in-place in descending order of highest to lowest priority.
+	// Pods at the start of the slice are less likely to be preempted.
+	OrderedPods func(eligible []*framework.PodInfo)
```
Will `OrderPods` be better since this returns a function instead of a sorted pod slice

nickbp replied:
Yes that sounds better, I wasn't super sure about the naming myself. Renamed here and throughout

nickbp committed:
Implement tests with example customizations, add direct constructor

nickbp committed:
rename OrderedPods -> OrderPods

nickbp committed:
Fix gofmt in default_preemption_test.go

macsko commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -65,7 +84,14 @@ func (pl *DefaultPreemption) Name() string {
+	return NewDefaultPreemption(ctx, dpArgs, fh, fts)
+}
+
+var _ frameworkruntime.PluginFactoryWithFts = New
```
This check is redundant. Code won't compile if the `New` doesn't match `frameworkruntime.PluginFactoryWithFts`

nickbp replied:
That's true, I'd added this as a "just in case" but the build would indeed fail elsewhere. Removed.

macsko commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -1472,6 +1458,279 @@ func TestSelectBestCandidate(t *testing.T) {
+			for _, p := range s.Victims().Pods {
+				gotPodNames = append(gotPodNames, p.Name)
+			}
+			if !reflect.DeepEqual(expectPodNames, gotPodNames) {
```
Please use `cmp.Diff` to compare

nickbp replied:
Done

macsko commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -1532,18 +1791,29 @@ func TestPodEligibleToPreemptOthers(t *testing.T) {
 			for _, n := range test.nodes {
 				nodes = append(nodes, st.MakeNode().Name(n).Obj())
 			}
+			var pods []runtime.Object
+			pods = append(pods, test.pod)
+			for _, pod := range test.pods {
+				pods = append(pods, pod)
+			}
```
```suggestion
			pods = append([]runtime.Object{test.pod}, test.pods...)
```

nickbp replied:
I'm getting a type error on this change:
```
default_preemption_test.go:1794:46: cannot use test.pods (variable of type []*"k8s.io/api/core/v1".Pod) as []"k8s.io/apimachinery/pkg/runtime".Object value in argument to append
```

Tried changing `test.pods` itself to be a `[]runtime.Object` but the `[]*v1.Pod` typing is needed further down when it's passed to `internalcache.NewSnapshot()`.

I'll leave it as-is for now but LMK if there's something else to try here.

macsko reviewed (commented):
Approach looks good

nickbp committed:
Remove unnecessary context and typecheck, switch to cmp.Diff

nickbp commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -65,7 +84,14 @@ func (pl *DefaultPreemption) Name() string {
+var _ frameworkruntime.PluginFactoryWithFts = New
+
+// NewDefaultPreemption initializes a new plugin and returns it. The plugin type is retained to allow modification.
+func NewDefaultPreemption(_ context.Context, dpArgs runtime.Object, fh framework.Handle, fts feature.Features) (*DefaultPreemption, error) {
```
Good point, done

nickbp commented:
At this point I'm not aware of anything else to be done here, let me know if I'm missing anything.

googs1025 commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -66,6 +84,11 @@ func (pl *DefaultPreemption) Name() string {
+}
+
+// NewDefaultPreemption initializes a new plugin and returns it. The plugin type is retained to allow modification.
+func NewDefaultPreemption(dpArgs runtime.Object, fh framework.Handle, fts feature.Features) (*DefaultPreemption, error) {
```
For me, this change will more flexibility, other developers can directly call `NewDefaultPreemption` and set EligiblePods and OrderPods by themselves. 👍 
```go
dp, _ := defaultpreemption.NewDefaultPreemption()
dp.EligiblePods = ...
dp.OrderPods = ...
```

googs1025 commented:
/test pull-kubernetes-e2e-gce

sanposhiho commented:
This week, the maintainers are pretty busy for the enhancement freeze. @googs1025 can you put a `/lgtm` when all things looking great to you and then I can make a second reviewing path next week.

googs1025 commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -1472,6 +1457,279 @@ func TestSelectBestCandidate(t *testing.T) {
+			if !ok {
+				t.Errorf("expect any node in %v, but got %v", tt.expected, s.Name())
+			}
+			gotPodNames := []string{}
+			for _, p := range s.Victims().Pods {
+				gotPodNames = append(gotPodNames, p.Name)
+			}
```
Do we care about the order here? Will CI fake appear because of the order?

nickbp replied:
The underlying evaluation is done in fixed order currently, so I wouldn't expect there to be any changes to the ordering across CI runs

nickbp replied:
But that said, LMK if it'd make sense to have this test case avoid expecting a particular ordering

googs1025 replied:
Got it, I just wanted to confirm if there is a CI failure issue due to the order. If you have confirmed it, I have no other questions about it.

dom4ha commented:
/cc

dom4ha commented:
The request to make the selection and ordering customizable is legitimate, but the interface should be as narrow as possible, to not block the future code refactoring.

For instance if we ever decide to put victim candidates into a heap, we'd need ordering function (less) that we can supply to the queue. The same refers to the filter. If we ever need to provide filtering function, we would not be able to use EligiblePods.

So please extract LessImportant and IsPodEligible or equivalent functions instead.

nickbp committed:
Update the customizations to operate on individual pods, for more flexibility later

nickbp committed:
Rename EligiblePod* to IsEligiblePod*

nickbp commented:
That makes sense. I've updated the customization funcs to both operate on individual pods as suggested. As a result this has also resulted in a smaller diff for `default_preemption.go` since the existing `for` loops can be retained as-is.

I also added an example test case/use case of customizing the preemption to be limited to specific nodes.

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -46,6 +46,16 @@ import (
 const Name = names.DefaultPreemption
 
+// IsEligiblePodFunc is a function which may be assigned to the DefaultPreemption plugin.
+// This function returns whether a given victim pod on the provided nodeInfo is eligible
```
```suggestion
// Implemenmtations should return whether a given victim pod on the provided nodeInfo is eligible
```

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -46,6 +46,16 @@ import (
+type IsEligiblePodFunc func(nodeInfo *framework.NodeInfo, victim *framework.PodInfo, preemptor *v1.Pod) bool
+
+// MoreImportantPodFunc is a function which may be assigned to the DefaultPreemption plugin.
+// This function returns true if the priority of the first pod is higher than the second pod.
```
```suggestion
// The implementations should return true if the first pod is more important than the second pod and the second one should be considered for preemption before the first one.
```

Overwritten behavior does not need to look into priorities

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -46,6 +46,16 @@ import (
+
+// MoreImportantPodFunc is a function which may be assigned to the DefaultPreemption plugin.
+// This function returns true if the priority of the first pod is higher than the second pod.
+// If the two pods are of equal priority then this will return false.
```
Please remove, especially since it's not a true for ordering function, which requires strict deterministic order to function properly.

nickbp replied:
Done

nickbp committed:
Apply suggestions from code review

Co-authored-by: Dominik Marciński <gmidon@gmail.com>

nickbp committed:
Remove line about equal priority, fix typo

macsko commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -66,6 +84,11 @@ func (pl *DefaultPreemption) Name() string {
 
 // New initializes a new plugin and returns it.
 func New(_ context.Context, dpArgs runtime.Object, fh framework.Handle, fts feature.Features) (framework.Plugin, error) {
+	return NewDefaultPreemption(dpArgs, fh, fts)
+}
+
+// NewDefaultPreemption initializes a new plugin and returns it. The plugin type is retained to allow modification.
+func NewDefaultPreemption(dpArgs runtime.Object, fh framework.Handle, fts feature.Features) (*DefaultPreemption, error) {
```
Couldn't we use generics in pkg/scheduler/framework/runtime/registry.go to keep the only one constructor here (that returns *DefaultPreemption)?

It should work if we change this:
https://github.com/kubernetes/kubernetes/blob/a18b4a8d97151a25c4d8088839d77e6ba482cac3/pkg/scheduler/framework/runtime/registry.go#L33-L42
To:
```go
// PluginFactoryWithFts is a function that builds a plugin with certain feature gates.
type PluginFactoryWithFts[T framework.Plugin] func(context.Context, runtime.Object, framework.Handle, plfeature.Features) (T, error)

// FactoryAdapter can be used to inject feature gates for a plugin that needs
// them when the caller expects the older PluginFactory method.
func FactoryAdapter[T framework.Plugin](fts plfeature.Features, withFts PluginFactoryWithFts[T]) PluginFactory {
	return func(ctx context.Context, plArgs runtime.Object, fh framework.Handle) (framework.Plugin, error) {
		return withFts(ctx, plArgs, fh, fts)
	}
}
```

nickbp replied:
I just tried this out and it seems to work! I was able to consolidate plugin construction to a single `New()` call, meanwhile not seeing issues elsewhere.

I'll put up another commit with the change, but can also revert/back that out if it doesn't make sense to do here?

nickbp committed:
Update factory to use generics, keep single New function

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -54,6 +63,15 @@ type DefaultPreemption struct {
+
+	// MoreImportantPod is used to sort eligible victims in-place in descending order of highest to
+	// lowest importance. Pods with higher importance are less likely to be preempted.
+	MoreImportantPod MoreImportantPodFunc
```
It's worth to mention the default behavior (priority + timestamp)

nickbp replied:
Done

dom4ha reviewed (commented):
Do we really want to allow completely overriding pod priorities, or only allow to filter out some of the lower priority pods (as eligible) and change the order only when two pods have the same priority?

nickbp commented:
> Do we really want to allow completely overriding pod priorities, or only allow to filter out some of the lower priority pods (as eligible) and change the order only when two pods have the same priority?

I'd rather not be too proscriptive about it - if anything having additional required filtering/sorting introduces the risk of blocking preemptions that the overrider wanted, or worse allowing preemptions that they didn't want.

nickbp committed:
Apply suggestions from code review

Co-authored-by: Dominik Marciński <gmidon@gmail.com>

nickbp committed:
Update comments: affinity info, default behavior, priority->importance

nickbp committed:
Clean up the affinity explanation, fix for new lint rule

dom4ha commented:
> > Do we really want to allow completely overriding pod priorities, or only allow to filter out some of the lower priority pods (as eligible) and change the order only when two pods have the same priority?
> 
> I'd rather not be too proscriptive about it - if anything having additional required filtering/sorting introduces the risk of blocking preemptions that the overrider wanted, or worse allowing preemptions that they didn't want.

I'm less concerned about the importance ordering function, but allowing preempting higher priority pods does not sound right. Do you have any example when it would be needed/desired?

nickbp commented:
I think the main use-case comes out of kube-scheduler using the same priority class value for two different independent decisions:
- whether to preempt (here)
- priority ordering of pending pods for placement

In our case we have 5 priority classes, where only one of those is eligible for preemption, and the other four aren't. Meanwhile the other four classes are divided in terms of their deployment priority - whether they should be at the front of the scheduling queue when there's capacity pressure. In our case, the preemptible priority class does happen to have the lowest priority value, but we effectively don't use that value for deciding preemptibility at all - it's only used for queue ordering.

dom4ha commented:
> I think the main use-case comes out of kube-scheduler using the same priority class value for two different independent decisions:
> 
> * whether to preempt (here)
> * priority ordering of pending pods for placement
> 
> In our case we have 5 priority classes, where only one of those is eligible for preemption, and the other four aren't. Meanwhile the other four classes are divided in terms of their deployment priority - whether they should be at the front of the scheduling queue when there's capacity pressure. In our case, the preemptible priority class does happen to have the lowest priority value, but we effectively don't use that value for deciding preemptibility at all - it's only used for queue ordering.

IIUC what you're saying, there is no need to preempt higher priority pods by lower priority (the case I'm concerned the most). If this is the case, can we change the function to allow only filtering out lower priority pods as not eligible for preemption instead of allowing to select any pod as eligible? The name of the function could stay the same, although the default behavior would be to not filter out anything.

nickbp committed:
Have IsEligiblePod be supplemental to priority check, update tests

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -1472,6 +1457,279 @@ func TestSelectBestCandidate(t *testing.T) {
+			}
+			offset, numCandidates := pl.GetOffsetAndNumCandidates(int32(len(nodeInfos)))
+			candidates, _, _ := pl.Evaluator.DryRunPreemption(ctx, state, tt.pod, nodeInfos, nil, offset, numCandidates)
+			s := pl.Evaluator.SelectCandidate(ctx, candidates)
```
`SelectCandidate` returns only the best candidate, but you need to iterate over all nodes and candidates.

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -1439,24 +1428,20 @@ func TestSelectBestCandidate(t *testing.T) {
-			if s == nil || len(s.Name()) == 0 {
+			candidates, _, _ := pl.Evaluator.DryRunPreemption(ctx, state, tt.pod, nodeInfos, nil, offset, numCandidates)
+			s := pl.Evaluator.SelectCandidate(ctx, candidates)
+			if len(tt.expected) == 0 {
```
I think it should be:

if len(tt.expected) != 1 {
  t.Fatalf("only one best node can be expected, got %v", tt.expected)
}

bsalamat commented:
This PR is more of a refactoring PR, plus the tests. It doesn’t change the logic or behavior of the scheduler by itself. It would help if the description of the PR was changed to reflect that. Right now, the description gives the impression that the new behavior of preemptible and non-preemptible priority classes are implemented.

I have further questions about the reasons behind the PR which is described in the description. I understand that in certain situations, eviction is not desired:
1. A high priority pod that should be scheduled ahead of other pods in the queue, but we don’t want this pod to preempt other running pods. We already have a solution for this. You can create a [non-preempting priority class](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class), by setting “preemptionPolicy: Never”.
2. A low priority pod that can wait to be scheduled, but once scheduled, we don’t want it to be preempted. We can create a pod disruption budget and set “minAvailable: 100%” (or “maxUnavailable: 0”).

I also see a potential corner case, which is less likely to happen, but can have a severe impact. In Kubernetes, we give the highest priorities to system pods. The assumption is that those critical system pods can always evict other pods to keep the system running. In the extreme case of many non-preemptible lower priority pods which have used all the cluster resources, the critical system pods could be deprived and cause an outage.

nickbp commented:
> This PR is more of a refactoring PR, plus the tests. It doesn’t change the logic or behavior of the scheduler by itself. It would help if the description of the PR was changed to reflect that. Right now, the description gives the impression that the new behavior of preemptible and non-preemptible priority classes are implemented.

I've updated the description to make it a little more apparent that this isn't changing behavior.

> I have further questions about the reasons behind the PR which is described in the description. I understand that in certain situations, eviction is not desired:
> 
> 1. A high priority pod that should be scheduled ahead of other pods in the queue, but we don’t want this pod to preempt other running pods. We already have a solution for this. You can create a [non-preempting priority class](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/#non-preempting-priority-class), by setting “preemptionPolicy: Never”.

Yes, that's not what we're doing. `preemptionPolicy: Never` blocks a given higher-priority pod from preempting any lower-priority pods, but doesn't block a given lower-priority pod from being preempted by higher-priority pods.

> 2. A low priority pod that can wait to be scheduled, but once scheduled, we don’t want it to be preempted. We can create a pod disruption budget and set “minAvailable: 100%” (or “maxUnavailable: 0”).

PDBs are used for defragging surplus capacity, at a time that replacement pods can be scheduled quickly. Meanwhile the same workloads/PDBs are not necessarily allowed to be preempted, because preemptions themselves imply a shortfall where replacement pods are less likely to be scheduled quickly. Due to this, allowing preemption for a given workload is treated as a very explicit opt-in.

> I also see a potential corner case, which is less likely to happen, but can have a severe impact. In Kubernetes, we give the highest priorities to system pods. The assumption is that those critical system pods can always evict other pods to keep the system running. In the extreme case of many non-preemptible lower priority pods which have used all the cluster resources, the critical system pods could be deprived and cause an outage.

That's a good point but it does vary by environment. I'd argue that someone customizing their kube-scheduler build/plugins at this level would need to be aware of the requirements of their environment. Such an advanced user would be able to customize system priority classes with `preemptionPolicy: Never` and end up in the same situation.

bsalamat commented:
I don't want to enforce my own ideas of how your clusters should be configured. I don't think I will ever configure my clusters to have low priority non-preemptible pods, because I think that's an anti-pattern, but I understand that others may have a different opinion and a reason to do so.

> Such an advanced user would be able to customize system priority classes with 'preemptionPolicy: Never' and end up in the same situation.

Those system priorities are predefined and users cannot change them. I am generally against giving users risky options. K8s is a complex systems and we all should do our best to make it hard for users to misconfigure their clusters.

All being said, I am not against this PR, given that it does not change the behavior of the scheduler and also the fact that only advanced users who create their own plugins would use the mechanism introduced here.

dom4ha commented:
I think the change itself is fine, since it allows having custom logic, but it does not influence the default behavior.

@sanposhiho @Huang-Wei @macsko and others, any objections?

@nickbp Note that there is a freeze today, but there are still small unresolved comments. Do you want it to make 1.33?

nickbp committed:
Have separate tests for custom selection vs ordering, add comments around system pod eligibility

nickbp reviewed (commented):
Sorry about the wait, have had some travel in the last week.
- Added comments to the filter function around ensuring that system pods are able to preempt
- Split the new tests into two separate ones for filtering vs ordering, and updated the filtering test to check all the candidates (was effectively just checking the first candidate)

nickbp commented:
OK, from giving everything a quick skim, I _think_ things are caught up at this point but LMK if I'd missed something...

nickbp committed:
gofmt

nickbp commented:
How's this looking at this point?

macsko commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:
```diff
@@ -1472,6 +1453,368 @@ func TestSelectBestCandidate(t *testing.T) {
+					gotVictims = append(gotVictims, p.Name)
+				}
+				if diff := cmp.Diff(expectVictims, gotVictims); diff != "" {
+					t.Errorf("expect victims %+v on node %s, but got pods %+v", expectVictims, selected.Name(), gotVictims)
```
If you use diff, always print it:
```suggestion
					t.Errorf("Unexpected victims on node %s (-want,+got):\n%s", selected.Name(), diff)
```

macsko commented:
> How's this looking at this point?

I like this approach, let's wait for others' opinions.

bsalamat commented:
I am not opposed to this PR either. I no longer have approval power though.
@dom4ha  and @Huang-Wei, please chime in.

nickbp committed:
Update pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go

Co-authored-by: Maciej Skoczeń <87243939+macsko@users.noreply.github.com>

nickbp commented:
/retest

nickbp commented:
FWIW the `pull-kubernetes-e2e-gce` test seems to be broken for unrelated reasons with initializing the environment. Meanwhile the only change since the last passed build was updating a unit test's failure output.

```
2025/04/22 05:40:42 main.go:326: Something went wrong: starting e2e cluster: error during ./hack/e2e-internal/e2e-up.sh: exit status 2
```

@sanposhiho @Huang-Wei @macsko what do you all think at this point?

haosdent commented:
/test pull-kubernetes-e2e-gce

sanposhiho commented:
Hmm, can you once rebase upon the latest master? We can ask the infra team if the rebase doesn't solve.

nickbp commented:
Just did a rebase to latest upstream, let's see how it goes :crossed_fingers:

nickbp commented:
Alrighty, CI is looking good again, thanks for the tip

dom4ha commented on pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:
```diff
@@ -46,6 +46,23 @@ import (
+// and the second one should be considered for preemption before the first one.
+// For performance reasons, the search for nodes eligible for preemption is done by omitting all
+// eligible victims from a node then checking whether the preemptor fits on the node without them,
+// before adding back victims that still fit with the preemptor.
```
```suggestion
// before adding back victims (starting from the most important) that still fit with the preemptor.
```

dom4ha commented:
/approve
/hold

@sanposhiho @macsko please have a look as well, especially to the change in runtime/registry.go to make the factory return the type. I don't see a problem with this change, but may be missing something.

nickbp committed:
Update pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

Co-authored-by: Dominik Marciński <gmidon@gmail.com>

nickbp commented on pkg/scheduler/framework/runtime/registry.go:
```diff
@@ -31,11 +31,11 @@ import (
 type PluginFactory = func(ctx context.Context, configuration runtime.Object, f framework.Handle) (framework.Plugin, error)
```
Gave this a few attempts just now and remembered that I think I'd tried this originally. Switching this to generics ends up causing typing problems around the `Registry` (aka `map[string]PluginFactory`) where we'd want everything to be a `framework.Plugin` anyway.

sanposhiho replied:
Okay, I guess there is a way, but no need to do in this PR anyways. I'll follow up in another PR.

sanposhiho reviewed (approved):
/lgtm
/approve

sanposhiho commented:
/retest

sanposhiho commented:
/release-note-none

macsko commented:
/unhold
